packages/presence-signal/src/createPresenceSwitchSignal.ts

```diff
@@ -0,0 +1,32 @@
+import { createSignal, createEffect, type Accessor } from "solid-js";
+import createPresenceSignal from "./createPresenceSignal";
+
+export default function createPresenceSwitchSignal<ItemType>(
+  item: Accessor<ItemType | undefined>,
+  opts: Parameters<typeof createPresenceSignal>[1],
+) {
+  const [mountedItem, setMountedItem] = createSignal(item());
+  const [shouldBeMounted, setShouldBeMounted] = createSignal(item() !== undefined);
+  const { isMounted, ...rest } = createPresenceSignal(shouldBeMounted, opts);
+
+  createEffect(() => {
+    if (mountedItem() !== item()) {
+      if (isMounted()) {
+        setShouldBeMounted(false);
+      } else if (item() !== undefined) {
+        setMountedItem(() => item());
+        setShouldBeMounted(true);
+      }
+    } else if (item() === undefined) {
+      setShouldBeMounted(false);
+    } else if (item() !== undefined) {
+      setShouldBeMounted(true);
+    }
+  });
```

Comment:
There is awfully lot of state synchronization using `createEffect`. Couldn't this be expressed with `createMemo`?

Reply:
I like that idea and it makes sense, however, I want to get the current version fully tested _and then_ start making changes to the core logic and validate against previous working tests. My main concern is the fact that lines 17 & 18 run next to each other here and if they were split up, I could see some potential jankiness being triggered from it.

Reply:
So I just added tests and started looking back at this idea, and I don't think it's possible. There are two states being synced here and I believe their relationships with other states requires an effect to keep synced.

## The reason we need an effect for `mountedItem`

The first line checks if `mountedItem() !== item()`, which inherently means that `mountedItem`'s future values can't be derived without being able to reference it's past values. So this flow has to be able to hold and reference previously existing state for that.

## The reason we need an effect for `shouldBeMounted`

The underlying `createPresence` call relies on the `shouldBeMounted` state, however, `createPresence` returns `isMounted` which the `shouldBeMountedState` itself utilizes to derive it's upcoming values. This seems like a chicken & the egg problem, but it's really just that `shouldBeMounted` is the true intended future state after all transitions whereas `isMounted` is the currently rendered state.